### PR TITLE
[2905]: Add Karras pattern to discrete euler

### PR DESCRIPTION
**Pull request by nipunjindal**
Discussion: #2905, #2064

This pull request allows for the utilization of karras_sigmas in the Euler Discrete scheduler. Additionally, happy to get feedback as we refine a set of guidelines as we expand this functionality to all the schedulers. This update introduces two key components:

- The inclusion of karras-based sigma.
- The mapping of sigmas to the timestamp, which will be used as input for UNET.

Also, I apologize if I overstepped any boundaries, as I was working on #2064, I realized to better setup the guidelines with a simpler scheduler.

**Euler Discrete Test**
```
from diffusers import StableDiffusionPipeline, EulerDiscreteScheduler
import torch 

seed = 33

pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("mps")
pipe.scheduler = EulerDiscreteScheduler.from_config(pipe.scheduler.config, use_karras_sigmas=True)

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="mps").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20).images[0]

display(image)
```
image

**Compared with KDiffusion - sample_euler**
```
import torch
from diffusers import StableDiffusionKDiffusionPipeline

seed = 33

pipe = StableDiffusionKDiffusionPipeline.from_pretrained('CompVis/stable-diffusion-v1-4').to('mps')
pipe.set_scheduler('sample_euler')

prompt = "an astronaut riding a horse on mars"

generator = torch.Generator(device="mps").manual_seed(seed)
image = pipe(prompt, generator=generator, num_inference_steps=20, use_karras_sigmas=True).images[0]

display(image)
```
image

Tensor Comparison

EulerDiscreteScheduler - Sigmas
```
14.6146, 11.7254,  9.3402,  7.3836,  5.7894,  4.4998,  3.4648,  2.6408, 1.9909,  1.4832,  1.0908,  0.7909,  0.5647,  0.3964,  0.2730,  0.1842, 0.1213,  0.0779,  0.0485,  0.0292,  0.0000
```
EulerDiscreteScheduler - Timestep to UNET
```
999.0000, 961.7328, 921.0421, 876.3138, 826.7906, 771.5520, 709.5297, 639.6147, 560.9754, 473.7489, 380.1317, 285.3175, 197.0806, 123.3951, 69.2569,  34.6974,  15.4816,   5.9918,   1.7830,   0.0000
```

sample_euler - Sigmas
```
14.6146, 11.7254, 9.3402, 7.3836, 5.7894, 4.4998, 3.4648, 2.6408, 1.9909, 1.4832, 1.0908, 0.7909, 0.5647, 0.3964, 0.2730, 0.1842, 0.1213, 0.0779, 0.0485, 0.0292, 0.0000
```

sample_euler - Timestep to UNET
```
998.9999, 961.7327, 921.0419, 876.3136, 826.7905, 771.5519, 709.5297, 639.6146, 560.9752, 473.7487, 380.1315, 285.3175, 197.0805, 123.3950, 69.2569, 34.6974, 15.4816, 5.9918, 1.7829, 0.0, 0.0
```

### CODE REVIEW

**Comment by nipunjindal** [comment]
Hello @sayakpaul @patrickvonplaten,

Please review and provide your feedback.
Thanks so much!

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 1]
```diff
@@ -103,6 +103,10 @@ class EulerDiscreteScheduler(SchedulerMixin, ConfigMixin):
         interpolation_type (`str`, default `"linear"`, optional):
             interpolation type to compute intermediate sigmas for the scheduler denoising steps. Should be one of
             [`"linear"`, `"log_linear"`].
+        use_karras_sigmas (`bool`, *optional*, defaults to `False`):
+            Use karras sigmas. For example, specifying `sample_dpmpp_2m` to `set_scheduler` will be equivalent to
+            `DPM++2M` in stable-diffusion-webui. On top of that, setting this option to True will make it `DPM++2M
+            Karras`.
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_euler_discrete.py:109]: Could we also include the reference paper that introduced it?

https://github.com/huggingface/diffusers/pull/2874#discussion_r1154578720

Feel free to also include it in `src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion_k_diffusion.py`.

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 2]
```diff
@@ -206,6 +217,45 @@ def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.devic
...
+    # Copied from https://github.com/crowsonkb/k-diffusion/blob/686dbad0f39640ea25c8a8c6a6e56bb40eacefa2/k_diffusion/sampling.py#L17
+    def _convert_to_karras(self, in_sigmas: torch.FloatTensor) -> torch.FloatTensor:
+        """Constructs the noise schedule of Karras et al. (2022)."""
+
+        sigma_min: float = in_sigmas[-1].item()
+        sigma_max: float = in_sigmas[0].item()
+        print(sigma_min, sigma_max)
+
+        rho = 7.0
+        # ramp = torch.linspace(0, 1, self.num_inference_steps)
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_euler_discrete.py:252]: Probably needs to go?

File: tests/schedulers/test_scheduler_euler.py [hunk 3]
```diff
@@ -117,3 +117,30 @@ def test_full_loop_device(self):
...
+            sample = scheduler.scale_model_input(sample, t)
+
+            model_output = model(sample, t)
+
+            output = scheduler.step(model_output, t, sample, generator=generator)
+            sample = output.prev_sample
+
+        result_sum = torch.sum(torch.abs(sample))
+        result_mean = torch.mean(torch.abs(sample))
+        print(result_sum.item(), result_mean.item())
```
Comment (sayakpaul) [inline tests/schedulers/test_scheduler_euler.py:144]: Needs to go?

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 4]
```diff
@@ -206,6 +217,45 @@ def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.devic
...
+        t = t.reshape(sigma.shape)
+        return t
+
+    # Copied from https://github.com/crowsonkb/k-diffusion/blob/686dbad0f39640ea25c8a8c6a6e56bb40eacefa2/k_diffusion/sampling.py#L17
+    def _convert_to_karras(self, in_sigmas: torch.FloatTensor) -> torch.FloatTensor:
+        """Constructs the noise schedule of Karras et al. (2022)."""
+
+        sigma_min: float = in_sigmas[-1].item()
+        sigma_max: float = in_sigmas[0].item()
+        print(sigma_min, sigma_max)
```
Comment (sayakpaul) [inline src/diffusers/schedulers/scheduling_euler_discrete.py:249]: Needs to go.

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 5]
```diff
@@ -206,6 +217,45 @@ def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.devic
...
+        # transform interpolation to time range
+        t = (1 - w) * low_idx + w * high_idx
+        t = t.reshape(sigma.shape)
+        return t
+
+    # Copied from https://github.com/crowsonkb/k-diffusion/blob/686dbad0f39640ea25c8a8c6a6e56bb40eacefa2/k_diffusion/sampling.py#L17
+    def _convert_to_karras(self, in_sigmas: torch.FloatTensor) -> torch.FloatTensor:
+        """Constructs the noise schedule of Karras et al. (2022)."""
+
+        sigma_min: float = in_sigmas[-1].item()
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/schedulers/scheduling_euler_discrete.py:247]: I think since `sigma_min` and `sigma_max` are derived values (being derived from `in_sigmas`), I think it's fine to not register the, separately.

**Review by sayakpaul — APPROVED** [review]
What a clean PR! Well done 🔥 

I am wondering if it would make sense to add this utility to 

https://github.com/huggingface/diffusers/blob/8c530fc2f6a76a2aefb6b285dce6df1675092ac6/src/diffusers/schedulers/scheduling_utils.py#L62

so that all the compatible schedulers can use Karras sigmas without having to duplicate code. 

WDYT?

**Comment by sayakpaul** [comment]
Also, I am wondering if we could somehow find a way to reuse `get_sigmas_karras()` from `k_diffusion` as done in https://github.com/huggingface/diffusers/pull/2874/files#diff-aea44e0665dd1127ca6b341af8cc731a4f36817c19553ddb559da2a919d1d267R20 ?

File: src/diffusers/schedulers/scheduling_euler_discrete.py (hunk 1 above)
Comment (nipunjindal) [reply src/diffusers/schedulers/scheduling_euler_discrete.py]: Fixed.

File: src/diffusers/schedulers/scheduling_euler_discrete.py (hunk 2 above)
Comment (nipunjindal) [reply src/diffusers/schedulers/scheduling_euler_discrete.py]: Removed the section.

File: tests/schedulers/test_scheduler_euler.py (hunk 3 above)
Comment (nipunjindal) [reply tests/schedulers/test_scheduler_euler.py]: Removed the section.

File: src/diffusers/schedulers/scheduling_euler_discrete.py (hunk 4 above)
Comment (nipunjindal) [reply src/diffusers/schedulers/scheduling_euler_discrete.py]: Removed the section.

**Comment by nipunjindal** [comment]
> Also, I am wondering if we could somehow find a way to reuse `get_sigmas_karras()` from `k_diffusion` as done in https://github.com/huggingface/diffusers/pull/2874/files#diff-aea44e0665dd1127ca6b341af8cc731a4f36817c19553ddb559da2a919d1d267R20 ?

Good point, utilized this function.

**Comment by nipunjindal** [comment]
> What a clean PR! Well done 🔥
> 
> I am wondering if it would make sense to add this utility to
> 
> https://github.com/huggingface/diffusers/blob/8c530fc2f6a76a2aefb6b285dce6df1675092ac6/src/diffusers/schedulers/scheduling_utils.py#L62
> 
> so that all the compatible schedulers can use Karras sigmas without having to duplicate code.
> 
> WDYT?

I am in favor of moving it to the base so that it becomes a few line change hopefully for other schedulers.

**Comment by nipunjindal** [comment]
Thank you for the review @sayakpaul, incorporated your feedback.

**Review by sayakpaul — APPROVED** [review]
Thanks for these changes. Looks really clean. 

@patrickvonplaten WDYT?

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 6]
```diff
@@ -18,6 +18,7 @@
 
 import numpy as np
 import torch
+from k_diffusion.sampling import get_sigmas_karras
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_euler_discrete.py:21]: We can import from the `k_diffusion` library here sadly. Could you just copy-paste the function here: https://github.com/crowsonkb/k-diffusion/blob/b43db16749d51055f813255eea2fdf1def801919/k_diffusion/sampling.py#L17 into the scheduler (ideally with a note that it's copied from the library to correctly cite)

File: src/diffusers/schedulers/scheduling_utils.py [hunk 7]
```diff
@@ -174,3 +175,27 @@ def _get_compatibles(cls):
             getattr(diffusers_library, c) for c in compatible_classes_str if hasattr(diffusers_library, c)
         ]
         return compatible_classes
+
+    @classmethod
+    def _sigma_to_t(self, sigma, log_sigmas):
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_utils.py:180]: Ok for me!

**Comment by patrickvonplaten** [comment]
> Also, I am wondering if we could somehow find a way to reuse `get_sigmas_karras()` from `k_diffusion` as done in https://github.com/huggingface/diffusers/pull/2874/files#diff-aea44e0665dd1127ca6b341af8cc731a4f36817c19553ddb559da2a919d1d267R20 ?

`k_diffusion` is not a hard dependency of `diffusers`, so we cannot import from `k_diffusion` I think the best idea here to just copy-paste the function between the relevant `schedulers` with the `#Copied from ` statement

**Comment by patrickvonplaten** [comment]
@nipunjindal amazing works, it's super cool that you're getting identical results between KDiffusion and `diffusers` - amazing job! 

Sorry for the back and forth with the review here, but it'd be nice if you could actually not add a dependency to `k_diffusion` inside `diffusers-only` schedulers and also not add `_sigma_to_t` to the general `SchedulerMixin` but directly to the specific scheduler. We can use the # Copy-from mechanism to copy-paste it everywhere.

**Comment by nipunjindal** [comment]
@patrickvonplaten Thank you for the reviews! I have addressed the review comments and confirmed the results are intact.

File: src/diffusers/schedulers/scheduling_euler_discrete.py (hunk 6 above)
Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_euler_discrete.py]: Original commit lgtm

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 8]
```diff
@@ -206,6 +217,45 @@ def set_timesteps(self, num_inference_steps: int, device: Union[str, torch.devic
...
+        w = (low - log_sigma) / (low - high)
+        w = np.clip(w, 0, 1)
+
+        # transform interpolation to time range
+        t = (1 - w) * low_idx + w * high_idx
+        t = t.reshape(sigma.shape)
+        return t
+
+    # Copied from https://github.com/crowsonkb/k-diffusion/blob/686dbad0f39640ea25c8a8c6a6e56bb40eacefa2/k_diffusion/sampling.py#L17
+    def _convert_to_karras(self, in_sigmas: torch.FloatTensor) -> torch.FloatTensor:
```
[the remark this answers is not in the document]
Comment (nipunjindal) [reply src/diffusers/schedulers/scheduling_euler_discrete.py:244]: Thanks. Fixed.

**Comment by sayakpaul** [comment]
@nipunjindal sorry for the oversight on my part and that you had to go through some repetitive cycles here.

**Review by patrickvonplaten — APPROVED** [review]
PR is good to go for me - @sayakpaul what do you think?

**Review by pcuenca — COMMENTED** [review]
Great work! I only have questions about the docstring :)

File: src/diffusers/schedulers/scheduling_euler_discrete.py [hunk 9]
```diff
@@ -103,6 +103,10 @@ class EulerDiscreteScheduler(SchedulerMixin, ConfigMixin):
         interpolation_type (`str`, default `"linear"`, optional):
             interpolation type to compute intermediate sigmas for the scheduler denoising steps. Should be one of
             [`"linear"`, `"log_linear"`].
+        use_karras_sigmas (`bool`, *optional*, defaults to `False`):
+            Use karras sigmas. For example, specifying `sample_dpmpp_2m` to `set_scheduler` will be equivalent to
+            `DPM++2M` in stable-diffusion-webui. On top of that, setting this option to True will make it `DPM++2M
+            Karras`. Please see equation (5) https://arxiv.org/pdf/2206.00364.pdf for more details.
```
[the remark this answers is not in the document]
Comment (nipunjindal) [reply src/diffusers/schedulers/scheduling_euler_discrete.py]: Right, these sigmas can be used generically in other scheduler pipelines as well. Updated the READme.

[the remark this answers is not in the document]
Comment (pcuenca) [reply src/diffusers/schedulers/scheduling_euler_discrete.py]: Thanks for iterating 🙏 I love the new text!

**Review by pcuenca — APPROVED** [review]
Good to go imo!

**Comment by sayakpaul** [comment]
@nipunjindal we deeply appreciate this important PR from you. Really great work :)
